src/groupeddataframe/grouping.jl

```diff
@@ -846,21 +883,53 @@ check_aggregate(::typeof(length)) = Aggregate(length)
 # among the first rows for each group
 function fillfirst!(condf, outcol::AbstractVector, incol::AbstractVector,
                     gd::GroupedDataFrame; rev::Bool=false)
-    nfilled = 0
-    @inbounds for i in eachindex(outcol)
-        s = gd.starts[i]
-        offsets = rev ? (nrow(gd[i])-1:-1:0) : (0:nrow(gd[i])-1)
-        for j in offsets
-            x = incol[gd.idx[s+j]]
-            if !condf === nothing || condf(x)
-                outcol[i] = x
+    ngroups = gd.ngroups
+    # Use group indices if they have already been computed
+    if gd.idx !== nothing && condf === nothing
+        v = rev ? gd.ends : gd.starts
+        idx = gd.idx
+        @inbounds for i in 1:ngroups
+            outcol[i] = incol[idx[v[i]]]
+        end
+    elseif gd.idx !== nothing
+        nfilled = 0
+        @inbounds for i in eachindex(outcol)
+            s = gd.starts[i]
+            offsets = rev ? (nrow(gd[i])-1:-1:0) : (0:nrow(gd[i])-1)
+            for j in offsets
+                x = incol[gd.idx[s+j]]
+                if !condf === nothing || condf(x)
+                    outcol[i] = x
+                    nfilled += 1
+                    break
+                end
+            end
+        end
+        if nfilled < length(outcol)
+            throw(ArgumentError("some groups contain only missing values"))
+        end
+    else # Finding first row is faster than computing all group indices
+        groups = gd.groups
+        if rev
+            r = length(groups):-1:1
+        else
+            r = 1:length(groups)
+        end
+        filled = fill(false, gd.ngroups)
```

Comment:
Why in this file in general do we use `fill(false, some_size` instead of `falses(some_size)`. I know there is a performance trade-off here, but I thought that recently `BitArray` as consistently faster than `Vector{Bool}` have you maybe benchmarked in recently?

Reply:
Probably worth checking, yes.

Reply:
So `BitArray` is still significantly slower on Julia 1.3, though the difference is irregular:
```julia


function fillfirst!(condf, outcol::AbstractVector, incol::AbstractVector,
                    gd::GroupedDataFrame; rev::Bool=false)
    ngroups = gd.ngroups
    # Use group indices if they have already been computed
    groups = gd.groups
    if rev
        r = length(groups):-1:1
    else
        r = 1:length(groups)
    end
    filled = fill(false, gd.ngroups)
    nfilled = 0
    @inbounds for i in r
        gix = groups[i]
        x = incol[i]
        if gix > 0 && (condf === nothing || condf(x)) && !filled[gix]
            filled[gix] = true
            outcol[gix] = x
            nfilled += 1
            nfilled == ngroups && break
        end
    end
    if nfilled < length(outcol)
        throw(ArgumentError("some groups contain only missing values"))
    end
    outcol
end


function fillfirst2!(condf, outcol::AbstractVector, incol::AbstractVector,
                     gd::GroupedDataFrame; rev::Bool=false)
    ngroups = gd.ngroups
    # Use group indices if they have already been computed
    groups = gd.groups
    if rev
        r = length(groups):-1:1
    else
        r = 1:length(groups)
    end
    filled = falses(gd.ngroups)
    nfilled = 0
    @inbounds for i in r
        gix = groups[i]
        x = incol[i]
        if gix > 0 && (condf === nothing || condf(x)) && !filled[gix]
            filled[gix] = true
            outcol[gix] = x
            nfilled += 1
            nfilled == ngroups && break
        end
    end
    if nfilled < length(outcol)
        throw(ArgumentError("some groups contain only missing values"))
    end
    outcol
end


using BenchmarkTools
n = 100_000
for k in (10, 100, 1000, 10_000)
    println("$k groups")
    df = DataFrame(x=rand(1:k, n), y=rand(n));
    gd = groupby(df, :x);
    out = Vector{Float64}(undef, k)
    in = Vector{Float64}(undef, n)
    @btime fillfirst!(nothing, $out, $in, $gd);
    @btime fillfirst2!(nothing, $out, $in, $gd);
end

10 groups
  70.659 ns (1 allocation: 96 bytes)
  94.355 ns (2 allocations: 128 bytes)
100 groups
  657.304 ns (1 allocation: 192 bytes)
  1.175 μs (2 allocations: 128 bytes)
1000 groups
  8.644 μs (1 allocation: 1.06 KiB)
  14.491 μs (2 allocations: 240 bytes)
10000 groups
  140.208 μs (1 allocation: 9.94 KiB)
  198.912 μs (2 allocations: 1.36 KiB)
```

Reply:
Interesting. Thank you for checking this!